aten/src/ATen/BatchedFallback.cpp

```diff
@@ -0,0 +1,142 @@
+  result.reserve(sizes.size());
+  for (auto it = sizes.rbegin(); it != sizes.rend(); it++) {
+    auto remainder = linear_idx % *it;
+    result.push_back(remainder);
+    linear_idx -= remainder;
+    linear_idx /= *it;
+  }
+  std::reverse(std::begin(result), std::end(result));
+  return result;
+}
+
+static bool areAllReturnsTensors(const FunctionSchema& schema) {
+  return std::all_of(
+      schema.returns().begin(),
+      schema.returns().end(),
+      [] (const Argument& arg) { return arg.type() == TensorType::get(); });
+}
+
+static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
+  for (const auto& arg : schema.arguments()) {
+    std::cout << arg.type() << std::endl;
+  }
+  return std::any_of(
+      schema.arguments().begin(),
+      schema.arguments().end(),
+      [] (const Argument& arg) { return arg.type()->isSubtypeOf(ListType::ofTensors()); });
+}
+
+void batchedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+  const auto& schema = op.schema();
+  auto num_returns = op.schema().returns().size();
+  TORCH_CHECK(!schema.is_mutable() && !schema.hasAnyAliasInfo(),
+              "Batching rule not implemented for ", schema, "; ",
+              "the fallback path doesn't work on in-place or view ops.");
+  TORCH_CHECK(areAllReturnsTensors(schema) && !areAnyArgumentsTensorList(schema),
+              "Batching rule not implemented for ", op.schema(), ". ",
+              "We could not generate a fallback.");
+  TORCH_CHECK(num_returns == 1,
+              "Batching rule not implemented for ", op.schema(), ". ",
+              "We do not yet support operations with multiple returns.");
+  TORCH_WARN("Batching rule not implemented for ", op.schema(), " falling back "
+             "to slow (for loop and stack) implementation");
+
+  // Figure out which arguments are BatchedTensor. Save them to a vector.
+  // For each BatchedTensor, also record what position of `stack` they came from.
+  std::vector<Tensor> batched_tensor_inputs;
+  VmapDimVector batched_tensor_inputs_position;
+  for (int64_t idx = 0; idx < stack->size(); ++idx) {
+    const auto& ivalue = (*stack)[idx];
+    if (!ivalue.isTensor()) {
+      continue;
+    }
+    const auto& tensor = ivalue.toTensor();
+    if (!tensor.defined()) {
+      continue;
+    }
+    const auto* batched = maybeGetBatched(tensor);
+    if (!batched) {
+      continue;
+    }
+    batched_tensor_inputs.push_back(tensor);
+    batched_tensor_inputs_position.push_back(idx);
+  }
+  TORCH_INTERNAL_ASSERT(batched_tensor_inputs.size() > 0);
+
+  // MultiBatchVmapTransform the BatchedTensor arguments. This returns
+  // VmapPhysicalViews that contain all of the batch dimensions.
+  const auto input_physical_views = MultiBatchVmapTransform::logicalToPhysical(
+      batched_tensor_inputs);
+
+  // Compute the total number of batches
+  auto num_batch_dims = input_physical_views.front().numBatchDims();
```

Comment:
Hmm, it's a little bit of a smell that we arbitrarily read out the first view for the number of batch dims, when in fact all of the views are guaranteed to have the same value (right?) Not sure if this is actionable though.

Reply:
Yes, all of the views have the same `numBatchDims()`. I'll think about if we can fix this somehow but my initial thought is that we might have to change the VmapTransforms API